src/io/memory_block_io.h

```diff
@@ -0,0 +1,218 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#if defined(ENABLE_SSE)
+#include <xmmintrin.h>  //todo
+#endif
+
+#include <algorithm>
+#include <cmath>
+#include <cstring>
+#include <nlohmann/json.hpp>
+#include <stdexcept>
+#include <vector>
+
+#include "basic_io.h"
+#include "common.h"
+#include "index/index_common_param.h"
+#include "vsag/allocator.h"
+#include "vsag/constants.h"
+
+namespace vsag {
+
+class MemoryBlockIO : public BasicIO<MemoryBlockIO> {
+public:
+    explicit MemoryBlockIO(Allocator* allocator, uint64_t block_size = DEFAULT_BLOCK_SIZE)
+        : block_size_(block_size), allocator_(allocator) {
+    }
+
+    MemoryBlockIO(const nlohmann::json& io_obj, const IndexCommonParam& common_param)
+        : allocator_(common_param.allocator_) {
+        CHECK_ARGUMENT(
+            io_obj.contains(BLOCK_IO_BLOCK_SIZE_KEY),
+            fmt::format("memory block io parameters must contains {}", BLOCK_IO_BLOCK_SIZE_KEY));
+
+        this->block_size_ = io_obj[BLOCK_IO_BLOCK_SIZE_KEY];  // TODO(LHT): trans str to uint64_t
+    }
+
+    ~MemoryBlockIO() override {
+        for (auto* block : blocks_) {
+            allocator_->Deallocate(block);
+        }
+    }
+
+    inline void
+    WriteImpl(const uint8_t* data, uint64_t size, uint64_t offset);
+
+    inline bool
+    ReadImpl(uint8_t* data, uint64_t size, uint64_t offset) const;
+
+    [[nodiscard]] inline const uint8_t*
+    ReadImpl(uint64_t size, uint64_t offset) const;
+
+    inline bool
+    MultiReadImpl(uint8_t* datas, uint64_t* sizes, uint64_t* offsets, uint64_t count) const;
+
+    inline void
+    PrefetchImpl(uint64_t offset, uint64_t cache_line = 64);
+
+    inline void
+    SerializeImpl(StreamWriter& writer);
+
+    inline void
+    DeserializeImpl(StreamReader& reader);
+
+private:
+    [[nodiscard]] inline bool
+    checkValidOffset(uint64_t size) const {
+        return size <= blocks_.size() * block_size_;
+    }
+
+    inline void
+    checkAndRealloc(uint64_t size);
+
+    [[nodiscard]] inline const uint8_t*
+    getDataPtr(uint64_t offset) const {
+        auto block_no = offset / block_size_;
+        auto block_off = offset % block_size_;
+        return blocks_[block_no] + block_off;
+    }
+
+    [[nodiscard]] inline bool
+    checkInOneBlock(uint64_t off1, uint64_t off2) const {
+        return (off1 / block_size_) == (off2 / block_size_);
+    }
+
+private:
+    uint64_t block_size_{DEFAULT_BLOCK_SIZE};
+
+    std::vector<uint8_t*> blocks_{};
+
+    Allocator* const allocator_{nullptr};
```

Comment:
`Allocator* const` to `const Allocator*`

Reply:
I think T* const is better